File: src/main/java/io/zeitwert/ddd/property/model/impl/PropertyHandler.java
package io.zeitwert.ddd.property.model.impl;

import io.zeitwert.ddd.aggregate.model.Aggregate;
import io.zeitwert.ddd.enums.model.Enumerated;
import io.zeitwert.ddd.property.model.EntityWithProperties;
import io.zeitwert.ddd.property.model.EnumProperty;
import io.zeitwert.ddd.property.model.EnumSetProperty;
import io.zeitwert.ddd.property.model.PartListProperty;
import io.zeitwert.ddd.property.model.Property;
import io.zeitwert.ddd.property.model.ReferenceProperty;
import io.zeitwert.ddd.property.model.ReferenceSetProperty;
import io.zeitwert.ddd.property.model.SimpleProperty;
import javassist.util.proxy.MethodHandler;

import java.lang.reflect.Field;
import java.lang.reflect.Method;

public class PropertyHandler implements MethodHandler {

	public static final PropertyHandler INSTANCE = new PropertyHandler();

	@Override
	@SuppressWarnings("unchecked")
	public Object invoke(Object self, Method m, Method proceed, Object[] args) throws Throwable {
		String methodName = m.getName();
		try {
			if (this.isCollectionApi(methodName, args)) {
				String fieldName = this.getCollectionFieldName(methodName, args);
				Property<?> property = this.getProperty(self, fieldName);
				if (property instanceof PartListProperty) {
					if (args.length == 0) {
						if (methodName.startsWith("get") && methodName.endsWith("Count")) {
							return ((PartListProperty<?>) property).getPartCount();
						} else if (methodName.startsWith("get") && methodName.endsWith("List")) {
							return ((PartListProperty<?>) property).getParts();
						} else if (methodName.startsWith("clear") && methodName.endsWith("List")) {
							((PartListProperty<?>) property).clearParts();
							return null;
						} else if (methodName.startsWith("add")) {
							return ((PartListProperty<?>) property).addPart();
						}
					} else if (args.length == 1) {
						if (methodName.startsWith("get") && methodName.endsWith("ById")) {
							return ((PartListProperty<?>) property).getPartById((Integer) args[0]);
						} else if (methodName.startsWith("get")) {
							return ((PartListProperty<?>) property).getPart((Integer) args[0]);
						} else if (methodName.startsWith("remove")) {
							((PartListProperty<?>) property).removePart((Integer) args[0]);
							return null;
						}
					}
				} else if (property instanceof EnumSetProperty) {
					if (args.length == 0) {
						if (methodName.startsWith("get") && methodName.endsWith("Count")) {
							return ((EnumSetProperty<?>) property).getItems().size();
						} else if (methodName.startsWith("get") && methodName.endsWith("Set")) {
							return ((EnumSetProperty<?>) property).getItems();
						} else if (methodName.startsWith("clear") && methodName.endsWith("Set")) {
							((EnumSetProperty<?>) property).clearItems();
							return null;
						}
					} else if (args.length == 1) {
						if (methodName.startsWith("has")) {
							return ((EnumSetProperty<Enumerated>) property).hasItem((Enumerated) args[0]);
						} else if (methodName.startsWith("add")) {
							((EnumSetProperty<Enumerated>) property).addItem((Enumerated) args[0]);
							return null;
						} else if (methodName.startsWith("remove")) {
							((EnumSetProperty<Enumerated>) property).removeItem((Enumerated) args[0]);
							return null;
						}
					}
				} else if (property instanceof ReferenceSetProperty) {
					if (args.length == 0) {
						if (methodName.startsWith("get") && methodName.endsWith("Count")) {
							return ((ReferenceSetProperty<?>) property).getItems().size();
						} else if (methodName.startsWith("get") && methodName.endsWith("Set")) {
							return ((ReferenceSetProperty<?>) property).getItems();
						} else if (methodName.startsWith("clear") && methodName.endsWith("Set")) {
							((ReferenceSetProperty<?>) property).clearItems();
							return null;
						}
					} else if (args.length == 1) {
						if (methodName.startsWith("has")) {
							return ((ReferenceSetProperty<?>) property).hasItem((Integer) args[0]);
						} else if (methodName.startsWith("add")) {
							((ReferenceSetProperty<?>) property).addItem((Integer) args[0]);
							return null;
						} else if (methodName.startsWith("remove")) {
							((ReferenceSetProperty<?>) property).removeItem((Integer) args[0]);
							return null;
						}
					}
				}
			}
			Property<?> property = this.getProperty(self, this.getFieldName(methodName));
			if (this.isGetter(methodName, args)) {
				if (property instanceof EnumProperty) {
					return ((EnumProperty<Enumerated>) property).getValue();
				} else if (property instanceof ReferenceProperty) {
					if (m.getName().endsWith("Id")) {
						return ((ReferenceProperty<Aggregate>) property).getId();
					} else {
						return ((ReferenceProperty<Aggregate>) property).getValue();
					}
				} else if (property instanceof SimpleProperty) { // must be last
					return ((SimpleProperty<?>) property).getValue();
				}
			} else if (this.isSetter(m.getName(), args)) {
				if (property instanceof EnumProperty) {
					((EnumProperty<Enumerated>) property).setValue((Enumerated) args[0]);
				} else if (property instanceof ReferenceProperty) {
					if (m.getName().endsWith("Id")) {
						((ReferenceProperty<Aggregate>) property).setId((Integer) args[0]);
					} else {
						((ReferenceProperty<Aggregate>) property).setValue((Aggregate) args[0]);
					}
				} else if (property instanceof SimpleProperty) { // must be last
					((SimpleProperty<Object>) property).setValue(args[0]);
				}
			}
		} catch (NoSuchFieldException x) {
			throw new NoSuchMethodException(self.getClass().getSimpleName() + "." + methodName);
		}
		return null;
	}

	private Property<?> getProperty(Object obj, String fieldName)
			throws NoSuchFieldException, IllegalArgumentException, IllegalAccessException {
		EntityWithProperties entity = (EntityWithProperties) obj;
		Property<?> property = entity.getProperty(fieldName);
		if (property != null) {
			return property;
		}
		for (Class<?> c = obj.getClass().getSuperclass(); c != null; c = c.getSuperclass()) {
			try {
				Field field = c.getDeclaredField(fieldName);
				field.setAccessible(true);
				return (Property<?>) field.get(obj);
			} catch (NoSuchFieldException | SecurityException e) {
			}
		}
		if (fieldName.endsWith("Id")) {
			// try enumeration / object via field
			return this.getProperty(obj, fieldName.substring(0, fieldName.length() - 2));
		} else if (fieldName.endsWith("List")) {
			// try Set instead of List
			return this.getProperty(obj, fieldName.replace("List", "Set"));
		} else if (!fieldName.endsWith("_")) {
			// try with _ suffix
			return this.getProperty(obj, fieldName + "_");
		}
		throw new NoSuchFieldException(fieldName);
	}

	private String getFieldName(String methodName) throws NoSuchFieldException {
		if (methodName.startsWith("get")) {
			return this.getName(methodName.substring(3));
		} else if (methodName.startsWith("set")) {
			return this.getName(methodName.substring(3));
		}
		throw new NoSuchFieldException(methodName);
	}

	private boolean isCollectionApi(String methodName, Object[] args) {
		if (args.length == 0) {
			if (methodName.startsWith("get") && methodName.endsWith("Count")) {
				return true;
			} else if (methodName.startsWith("get") && (methodName.endsWith("List") || methodName.endsWith("Set"))) {
				return true;
			} else if (methodName.startsWith("clear") && (methodName.endsWith("List") || methodName.endsWith("Set"))) {
				return true;
			} else if (methodName.startsWith("add")) {
				return true;
			}
		} else if (args.length == 1) {
			if (methodName.startsWith("get") && methodName.endsWith("ById")) {
				return true;
			} else if (methodName.startsWith("has")) {
				return true;
			} else if (methodName.startsWith("get")) {
				return true;
			} else if (methodName.startsWith("add")) {
				return true;
			} else if (methodName.startsWith("remove")) {
				return true;
			}
		}
		return false;
	}

	private String getCollectionFieldName(String methodName, Object[] args) throws NoSuchFieldException {
		if (args.length == 0) {
			if (methodName.startsWith("get") && methodName.endsWith("Count")) {
				return this.getName(methodName.substring(3, methodName.length() - 5)) + "List";
			} else if (methodName.startsWith("get") && methodName.endsWith("List")) {
				return this.getName(methodName.substring(3, methodName.length() - 4)) + "List";
			} else if (methodName.startsWith("get") && methodName.endsWith("Set")) {
				return this.getName(methodName.substring(3, methodName.length() - 3)) + "Set";
			} else if (methodName.startsWith("clear") && methodName.endsWith("List")) {
				return this.getName(methodName.substring(5, methodName.length() - 4)) + "List";
			} else if (methodName.startsWith("clear") && methodName.endsWith("Set")) {
				return this.getName(methodName.substring(5, methodName.length() - 3)) + "Set";
			} else if (methodName.startsWith("add")) {
				return this.getName(methodName.substring(3)) + "List";
			}
		} else if (args.length == 1) {
			if (methodName.startsWith("get") && methodName.endsWith("ById")) {
				return this.getName(methodName.substring(3, methodName.length() - 4)) + "List";
			} else if (methodName.startsWith("has")) {
				return this.getName(methodName.substring(3)) + "Set";
			} else if (methodName.startsWith("get")) {
				return this.getName(methodName.substring(3)) + "List";
			} else if (methodName.startsWith("add")) {
				return this.getName(methodName.substring(3)) + "Set";
			} else if (methodName.startsWith("remove")) {
				return this.getName(methodName.substring(6)) + "List";
			}
		}
		throw new NoSuchFieldException(methodName);
	}

	private String getName(String methodName) throws NoSuchFieldException {
		return methodName.substring(0, 1).toLowerCase() + methodName.substring(1);
	}

	private boolean isGetter(String methodName, Object[] args) {
		return methodName.startsWith("get") && args.length == 0;
	}

	private boolean isSetter(String methodName, Object[] args) {
		return methodName.startsWith("set") && args.length == 1;
	}

}

File: fm-dddrive/src/main/java/io/dddrive/property/model/impl/PropertyHandler.kt
package io.dddrive.property.model.impl

import io.dddrive.ddd.model.Aggregate
import io.dddrive.ddd.model.Part
import io.dddrive.enums.model.Enumerated
import io.dddrive.property.model.AggregateReferenceProperty
import io.dddrive.property.model.BaseProperty
import io.dddrive.property.model.EntityWithProperties
import io.dddrive.property.model.EnumProperty
import io.dddrive.property.model.EnumSetProperty
import io.dddrive.property.model.PartListProperty
import io.dddrive.property.model.PartReferenceProperty
import io.dddrive.property.model.Property
import io.dddrive.property.model.ReferenceSetProperty
import javassist.util.proxy.MethodHandler
import java.lang.reflect.Method
import java.util.*

class PropertyHandler : MethodHandler {

	@Throws(Throwable::class)
	@Suppress("UNCHECKED_CAST")
	override fun invoke(
		self: Any,
		m: Method,
		proceed: Method?,
		args: Array<Any?>,
	): Any? {
		val methodName = m.name
		try {
			if (this.isCollectionApi(methodName, args)) {
				val fieldName = this.getCollectionFieldName(methodName, args)
				val property = this.getProperty(self, fieldName)
				if (property is PartListProperty<*>) {
					if (args.isEmpty()) {
						if (methodName.startsWith("get") && methodName.endsWith("Count")) {
							return property.partCount
						} else if (methodName.startsWith("get") && methodName.endsWith("List")) {
							return property.parts
						} else if (methodName.startsWith("clear") && methodName.endsWith("List")) {
							property.clearParts()
							return null
						} else if (methodName.startsWith("add")) {
							return property.addPart(null)
						}
					} else if (args.size == 1) {
						if (methodName.startsWith("get") && methodName.endsWith("ById")) {
							return property.getPartById((args[0] as Int?)!!)
						} else if (methodName.startsWith("get")) {
							return property.getPart((args[0] as Int?)!!)
						} else if (methodName.startsWith("remove")) {
							property.removePart((args[0] as Int?)!!)
							return null
						}
					}
				} else if (property is EnumSetProperty<*>) {
					if (args.isEmpty()) {
						if (methodName.startsWith("get") && methodName.endsWith("Count")) {
							return property.items.size
						} else if (methodName.startsWith("get") && methodName.endsWith("Set")) {
							return property.items
						} else if (methodName.startsWith("clear") && methodName.endsWith("Set")) {
							property.clearItems()
							return null
						}
					} else if (args.size == 1) {
						if (methodName.startsWith("has")) {
							return (property as EnumSetProperty<Enumerated>).hasItem((args[0] as Enumerated?)!!)
						} else if (methodName.startsWith("add")) {
							(property as EnumSetProperty<Enumerated>).addItem((args[0] as Enumerated?)!!)
							return null
						} else if (methodName.startsWith("remove")) {
							(property as EnumSetProperty<Enumerated>).removeItem((args[0] as Enumerated?)!!)
							return null
						}
					}
				} else if (property is ReferenceSetProperty<*>) {
					if (args.isEmpty()) {
						if (methodName.startsWith("get") && methodName.endsWith("Count")) {
							return property.items.size
						} else if (methodName.startsWith("get") && methodName.endsWith("Set")) {
							return property.items
						} else if (methodName.startsWith("clear") && methodName.endsWith("Set")) {
							property.clearItems()
							return null
						}
					} else if (args.size == 1) {
						if (methodName.startsWith("has")) {
							return property.hasItem(args[0]!!)
						} else if (methodName.startsWith("add")) {
							property.addItem(args[0]!!)
							return null
						} else if (methodName.startsWith("remove")) {
							property.removeItem(args[0]!!)
							return null
						}
					}
				}
			}
			var fieldName = this.getFieldName(methodName)
			val property = try {
				this.getProperty(self, fieldName)
			} catch (_: NoSuchFieldException) {
				try {
					// try boolean setXyz setter
					fieldName = "is" + fieldName.substring(0, 1).uppercase(Locale.getDefault()) + fieldName.substring(1)
					this.getProperty(self, fieldName)
				} catch (_: NoSuchFieldException) {
					throw NoSuchMethodException(self.javaClass.getSimpleName() + "." + methodName)
				}
			}
			if (this.isGetter(methodName, args)) {
				return if (property is EnumProperty<*>) {
					(property as EnumProperty<Enumerated>).value
				} else if (property is AggregateReferenceProperty<*>) {
					if (m.name.endsWith("Id")) {
						property.id
					} else {
						property.value
					}
				} else if (property is PartReferenceProperty<*>) {
					if (m.name.endsWith("Id")) {
						property.id
					} else {
						property.value
					}
				} else if (property is BaseProperty<*>) { // must be last
					property.value
				} else {
					throw NoSuchFieldException(property.name)
				}
			} else if (this.isSetter(m.name, args)) {
				if (property is EnumProperty<*>) {
					(property as EnumProperty<Enumerated>).value = args[0] as Enumerated?
				} else if (property is AggregateReferenceProperty<*>) {
					if (m.name.endsWith("Id")) {
						(property as AggregateReferenceProperty<Aggregate>).id = args[0]
					} else {
						(property as AggregateReferenceProperty<Aggregate>).value = args[0] as Aggregate?
					}
				} else if (property is PartReferenceProperty<*>) {
					if (m.name.endsWith("Id")) {
						property.id = args[0] as Int?
					} else {
						(property as PartReferenceProperty<Part<*>>).value = args[0] as Part<*>?
					}
				} else if (property is BaseProperty<*>) { // must be last
					(property as BaseProperty<Any?>).value = args[0]
				}
			}
		} catch (_: NoSuchFieldException) {
			throw NoSuchMethodException(self.javaClass.getSimpleName() + "." + methodName)
		}
		return null
	}

	@Throws(NoSuchFieldException::class, IllegalArgumentException::class)
	private fun getProperty(
		obj: Any?,
		fieldName: String,
	): Property<*> {
		val entity = obj as EntityWithProperties
		var property: Property<*>? = if (entity.hasProperty(fieldName)) entity.getProperty(fieldName, Any::class) else null
		if (property != null) {
			return property
		}
		// look for field with underscore prefix (e.g. _name), kotlin style
		property = if (entity.hasProperty("_$fieldName")) entity.getProperty("_$fieldName", Any::class) else null
		if (property != null) {
			return property
		}
		if (fieldName.endsWith("Id")) { // try enumeration / object via field
			return this.getProperty(obj, fieldName.substring(0, fieldName.length - 2))
		} else if (fieldName.endsWith("List")) { // try Set instead of List
			return this.getProperty(obj, fieldName.replace("List", "Set"))
		}
		throw NoSuchFieldException(fieldName)
	}

	@Throws(NoSuchFieldException::class)
	private fun getFieldName(methodName: String): String {
		if (methodName.startsWith("get")) {
			return this.getName(methodName.substring(3))
		} else if (methodName.startsWith("is")) {
			return methodName
		} else if (methodName.startsWith("set")) {
			return this.getName(methodName.substring(3))
		}
		throw NoSuchFieldException(methodName)
	}

	private fun isCollectionApi(
		methodName: String,
		args: Array<Any?>,
	): Boolean {
		if (args.size == 0) {
			if (methodName.startsWith("get") && methodName.endsWith("Count")) {
				return true
			} else if (methodName.startsWith("get") && (methodName.endsWith("List") || methodName.endsWith("Set"))) {
				return true
			} else if (methodName.startsWith("clear") && (methodName.endsWith("List") || methodName.endsWith("Set"))) {
				return true
			} else {
				return methodName.startsWith("add")
			}
		} else if (args.size == 1) {
			if (methodName.startsWith("get") && methodName.endsWith("ById")) {
				return true
			} else if (methodName.startsWith("has")) {
				return true
			} else if (methodName.startsWith("get")) {
				return true
			} else if (methodName.startsWith("add")) {
				return true
			} else {
				return methodName.startsWith("remove")
			}
		}
		return false
	}

	@Throws(NoSuchFieldException::class)
	private fun getCollectionFieldName(
		methodName: String,
		args: Array<Any?>,
	): String {
		if (args.size == 0) {
			if (methodName.startsWith("get") && methodName.endsWith("Count")) {
				return this.getName(methodName.substring(3, methodName.length - 5)) + "List"
			} else if (methodName.startsWith("get") && methodName.endsWith("List")) {
				return this.getName(methodName.substring(3, methodName.length - 4)) + "List"
			} else if (methodName.startsWith("get") && methodName.endsWith("Set")) {
				return this.getName(methodName.substring(3, methodName.length - 3)) + "Set"
			} else if (methodName.startsWith("clear") && methodName.endsWith("List")) {
				return this.getName(methodName.substring(5, methodName.length - 4)) + "List"
			} else if (methodName.startsWith("clear") && methodName.endsWith("Set")) {
				return this.getName(methodName.substring(5, methodName.length - 3)) + "Set"
			} else if (methodName.startsWith("add")) {
				return this.getName(methodName.substring(3)) + "List"
			}
		} else if (args.size == 1) {
			if (methodName.startsWith("get") && methodName.endsWith("ById")) {
				return this.getName(methodName.substring(3, methodName.length - 4)) + "List"
			} else if (methodName.startsWith("has")) {
				return this.getName(methodName.substring(3)) + "Set"
			} else if (methodName.startsWith("get")) {
				return this.getName(methodName.substring(3)) + "List"
			} else if (methodName.startsWith("add")) {
				return this.getName(methodName.substring(3)) + "Set"
			} else if (methodName.startsWith("remove")) {
				return this.getName(methodName.substring(6)) + "List"
			}
		}
		throw NoSuchFieldException(methodName)
	}

	private fun getName(methodName: String): String = methodName.substring(0, 1).lowercase(Locale.getDefault()) + methodName.substring(1)

	private fun isGetter(
		methodName: String,
		args: Array<Any?>,
	): Boolean = (methodName.startsWith("get") || methodName.startsWith("is")) && args.isEmpty()

	private fun isSetter(
		methodName: String,
		args: Array<Any?>,
	): Boolean = methodName.startsWith("set") && args.size == 1

	companion object {

		@JvmField
		val INSTANCE: PropertyHandler = PropertyHandler()
	}
}
